### PR TITLE
[Merged by Bors] - role based configuration

### PR DESCRIPTION
**Pull request by razvan**
Part of: #306 

This PR has been extracted from #320 which will be closed. The part that was left out is the actual configuration the of segment cache size. That will be implemented in a future PR and will require a new operator-rs release.

:green_circle: CI https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/druid-operator-it-custom/34/

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,8 @@ All notable changes to this project will be documented in this file.
 - Cpu and memory limits are now configurable ([#298]).
 - Stale resources are now deleted ([#310]).
 - Support Druid 24.0.0 ([#317]).
+- Refactor role configuration with per role structs like `BrokerConfig`, `HistoricalConfig`, etc ([#332])
+- Added `HistoricalStorge` and `DruidStorage` (as catch-all storage configuration) ([#332])
```
Comment (sbernauer) [inline CHANGELOG.md:13]:
```suggestion
- Added `HistoricalStorage` and `DruidStorage` (as catch-all storage configuration) ([#332])
```

File: deploy/config-spec/properties.yaml [hunk 2]
```diff
@@ -722,7 +722,7 @@ properties:
         type: "string"
       defaultValues:
         - fromVersion: "0.0.0"
-          value: "[{\"path\":\"/stackable/var/druid/segment-cache\",\"maxSize\":\"300g\"}]"
+          value: "[{\"path\":\"/stackable/var/druid/segment-cache\",\"maxSize\":\"900m\"}]"
```
Comment (sbernauer) [inline deploy/config-spec/properties.yaml:725]: Actually i would prefer not to change this, as this reduces the amount you can possible query on a historical.
I have made the observation that only data residing on local disk cache can be queried

File: rust/crd/src/resource.rs [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+    commons::resources::{CpuLimits, MemoryLimits, NoRuntimeLimits, Resources},
+    config::merge::Merge,
+    k8s_openapi::{
+        api::core::v1::ResourceRequirements, apimachinery::pkg::api::resource::Quantity,
+    },
+};
+use strum::{EnumDiscriminants, IntoStaticStr};
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum RoleResourceEnum {
```
Comment (sbernauer) [inline rust/crd/src/resource.rs:14]: Any particular reason for the Enum suffix? Reminds me of the Hungarian notation ^^

File: rust/crd/src/resource.rs [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+    first: Option<&mut RoleResourceEnum>,
+    second: Option<&mut RoleResourceEnum>,
+    third: Option<&mut RoleResourceEnum>,
+) -> Result<RoleResourceEnum, Error> {
+    let mut some = [first, second, third]
+        .into_iter()
+        .flatten()
+        .collect::<Vec<&mut RoleResourceEnum>>();
+
+    match some.len() {
```
Comment (sbernauer) [inline rust/crd/src/resource.rs:56]: There might be more clever ways but fine for me for 3 entries

Comment (razvan) [reply rust/crd/src/resource.rs:56]: If you can find a clever way that's also the same amount of lines of code or even less, I'm interested :)

File: rust/crd/src/resource.rs (hunk 3 above)
Comment (razvan) [reply rust/crd/src/resource.rs:14]: It's a leftover from trying different things out. I will rename it. Also I liked the Hungarian notation :)

File: deploy/config-spec/properties.yaml (hunk 2 above)
Comment (razvan) [reply deploy/config-spec/properties.yaml]: Reverted.

File: rust/crd/src/resource.rs (hunk 4 above)
Comment (sbernauer) [reply rust/crd/src/resource.rs:56]: I pushed a proposal to the `role-based-configuration-review` branch ;)
Might be something for operator-rs

**Review by sbernauer — APPROVED** [review]
LGTM % `role-based-configuration-review` proposal

**Comment by razvan** [comment]
bors merge

**Pull request closed without merging**